### PR TITLE
[#63] Interpret non-existing service directory as no existing services

### LINKED ISSUE #63: Service list fails when service directory does not exist.
## Required information

The code from the iceoryx2 discovery example will fail when the service directory does not exist, this is unexpected.

It should be interpreted as no active services.

### PR DESCRIPTION
**Pull request by elfenpiff**
## Pre-Review Checklist for the PR Author

1. [x] Add sensible notes for the reviewer
1. [x] PR title is short, expressive and meaningful
1. [x] Relevant issues are linked
1. [x] Every source code file has a copyright header with `SPDX-License-Identifier: Apache-2.0 OR MIT`
1. [x] Branch follows the naming format (`iox2-123-introduce-posix-ipc-example`)
1. [x] Commits messages are according to this [guideline][commit-guidelines]
    - [x] Commit messages have the issue ID (`[#123] Add posix ipc example`)
    - [x] Commit author matches [Eclipse Contributor Agreement][eca] (and ECA is signed)
1. [x] Tests follow the [best practice for testing][testing]
1. [x] Changelog updated [in the unreleased section][changelog] including API breaking changes
1. [x] Assign PR to reviewer
1. [x] All checks have passed (except `task-list-completed`)

[commit-guidelines]: https://tbaggery.com/2008/04/19/a-note-about-git-commit-messages.html
[eca]: http://www.eclipse.org/legal/ECA.php
[testing]: https://github.com/eclipse-iceoryx/iceoryx/blob/master/doc/website/concepts/best-practice-for-testing.md
[changelog]: https://github.com/eclipse-iceoryx/iceoryx2/blob/main/doc/release-notes/iceoryx2-unreleased.md

## Checklist for the PR Reviewer

- [x] Commits are properly organized and messages are according to the guideline
- [x] Unit tests have been written for new behavior
- [x] Public API is documented
- [x] PR title describes the changes

## Post-review Checklist for the PR Author

1. [x] All open points are addressed and tracked via issues

## References

Use either 'Closes #123' or 'Relates to #123' to reference the corresponding issue.

Closes #63

### CODE REVIEW

File: iceoryx2-cal/src/static_storage/file.rs [hunk 1]
```diff
@@ -213,10 +214,20 @@ impl crate::named_concept::NamedConceptMgmt for Storage {
...
-            unmatched NamedConceptListError::InternalError,
-            "{} due to a failure while reading the storage directory (\"{}\").", msg, config.path);
+        let directory = match Directory::new(&config.path) {
+            Ok(directory) => directory,
+            Err(DirectoryOpenError::InsufficientPermissions) => {
+                debug!(from origin, "{} due to insufficient permissions to read the storage directory.", msg);
+                return Err(NamedConceptListError::InsufficientPermissions);
+            }
+            Err(DirectoryOpenError::DoesNotExist) => {
+                return Ok(vec![]);
```
Comment (elBoberido) [inline iceoryx2-cal/src/static_storage/file.rs:224]: nitpick: the macro is not necessary and a `Ok(Vec::new())` would also work

File: iceoryx2-cal/src/static_storage/file.rs [hunk 2]
```diff
@@ -213,10 +214,20 @@ impl crate::named_concept::NamedConceptMgmt for Storage {
...
+            }
+            Err(v) => {
+                debug!(from origin, "{} due to failure ({:?}) while reading the storage directory (\"{}\").", msg, v, config.path);
+                return Err(NamedConceptListError::InternalError);
+            }
+        };
```
Comment (elBoberido) [inline iceoryx2-cal/src/static_storage/file.rs:231]: Btw, the `for entry in &entries` loop could be rewritten with iterators like this
```rs
let result = entries
    .iter()
    .filter(|entry| {
        let metadata = entry.metadata();
        metadata.file_type() == FileType::File && metadata.permission() == FINAL_PERMISSIONS
    })
    .filter_map(|entry| config.extract_name_from_file(entry.name()))
    .collect();
```

**Review by elBoberido — APPROVED** [review]
Can be merged as is but if you like, there is a suggestion to remove a for loop

File: iceoryx2-cal/src/static_storage/file.rs (hunk 1 above)
Comment (elfenpiff) [reply iceoryx2-cal/src/static_storage/file.rs:224]: @elBoberido are you sure, the other suggestion is more to type? Why would it be better?

Comment (elBoberido) [reply iceoryx2-cal/src/static_storage/file.rs:224]: Because `()` is easier to reach on my keyboard than `[]` ;)

**Review by elBoberido — APPROVED** [review]

**Pull request merged**